### PR TITLE
`when` with limited concurrently pending promises.

### PR DESCRIPTION
**Pull request by shergin**
The second approach for #444.

### CODE REVIEW

File: Sources/when.swift [hunk 1]
```diff
@@ -86,3 +86,88 @@ public func when<U, V>(pu: Promise<U>, _ pv: Promise<V>) -> Promise<(U, V)> {
...
+    var generator = promiseGenerator
+
+    return Promise { fulfill, reject in
+        var pendingPromises = 0
+        var values: [T?] = []
+        var rejected = false
+        var counter = 0
+
+        func extendArraySize(size: Int) {
+            if size > values.capacity {
```
Comment (shergin) [inline Sources/when.swift]: I just realized that probably I should not do it because standard library should extend capacity automatically.

**Comment by mxcl** [comment]
Looks great. Nice use of `AnyGenerator`. I approve of using the stdlib.

**Comment by shergin** [comment]
Cool!
So, I will add some more tests and we good to go (merge)?

**Comment by nathanhosselton** [comment]
@shergin I would say so.

**Comment by shergin** [comment]
I have just added several tests for it.
What do you think?

**Comment by mxcl** [comment]
Great, appreciated :)

The documentation for this function is a copy and paste from the other when. Can we make it describe what is different?

Running all PMKTests crashes for me:

screen shot 2016-07-16 at 11 13 47 am

Also, this is not thread safe.

**Comment by shergin** [comment]
Ough! Thanks! But I thought it is not finished yet. ) I am on vacation right now so I have not time to work on it. Sorry.
I have a fix for the crash that was shown on screenshot and I would love to discuss about thread-safety. I will open new PR for that. )

PS:
Max, nice avatar!

**Comment by mxcl** [comment]
I fixed the issues in subsequent commits, though… you should review!

**Comment by shergin** [comment]
@mxcl Awesome! Your approach is much better! Thank you, Max!

**Comment by nobre84** [comment]
Hey! This is a great under documented feature! Any reason why it isn't available for when(resolved:) as well ?

**Comment by mxcl** [comment]
No reason, PR welcome!

**Pull request merged**